warehouse/oidc/views.py

```diff
@@ -0,0 +1,119 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.events.tags import EventTag
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    # `json_body` can return any valid top-level JSON type, so we have
+    # to make sure we're actually receiving a dictionary.
+    if not isinstance(body, dict):
+        return _invalid(
+            errors=[
+                {
+                    "code": "invalid-payload",
+                    "description": "payload is not a JSON dictionary",
+                }
+            ]
+        )
+
+    unverified_jwt = body.get("token")
+    if unverified_jwt is None:
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is missing"}]
+        )
+
+    if not isinstance(unverified_jwt, str):
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is not a string"}]
+        )
+
+    # For the time being, GitHub is our only OIDC provider.
+    # In the future, this should locate the correct service based on an
+    # identifier in the request body.
+    oidc_service = request.find_service(IOIDCProviderService, name="github")
+    provider = oidc_service.find_provider(unverified_jwt)
+    if not provider:
+        return _invalid(
+            errors=[
+                {"code": "invalid-token", "description": "malformed or invalid token"}
+            ]
+        )
+
+    # At this point, we've verified that the given JWT is valid for the given
+    # project. All we need to do is mint a new token.
+    # NOTE: For OIDC-minted API tokens, the Macaroon's description string
+    # is purely an implementation detail and is not displayed to the user.
```

Comment:
Flagging for visibility: the other option here is to relax the `Macaroon.description` column and allow it to be nullable.

Reply:
I'm thinking it might be more helpful for this to be the `provider_url`.

Reply:
Whoops, forgot to get this with the last changeset. Fixing now!

Reply:
667c9fc5